### PR TITLE
[Jetpack] Enable Domains purchases for wpcom sites in adhoc builds

### PR DESCRIPTION
**Pull request by frosty**
This PR enables the Domains dashboard and purchasing flow for the Jetpack app, only for wpcom sites and only for adhoc builds.

Requires associated WordPressKit PR: https://github.com/wordpress-mobile/WordPressKit-iOS/pull/497

|   |   |   |   |
|---|---|---|---|
| Simulator Screen Shot - iPhone 13 Pro - 2022-04-22 at 13 54 56 | Simulator Screen Shot - iPhone 13 Pro - 2022-04-22 at 13 54 00 | Simulator Screen Shot - iPhone 13 Pro - 2022-04-22 at 13 54 08 | Simulator Screen Shot - iPhone 13 Pro - 2022-04-22 at 13 54 45 |

**To test**

- Build and run the Jetpack target
- View the Menu section of My Site for a number of sites – try to verify WordPress.com sites, self-hosted non-Jetpack and self-hosted Jetpack.
- For WordPress.com sites, you should see the Domains item in the My Site menu, provided that you have site management permissions.
- Navigate into Domains when it's present and ensure you can see the site's current domain listed.
- Attempt the entire domains purchase flow. It's recommended to use the sandbox store – for full information about setting this up, read #p2-pbArwn-2GN. You should be able to successfully complete the flow, and the Domains Dashboard should update and reflect the new domain.
- Also verify that the Domains section isn't visible in the WordPress app.

## Regression Notes

1. Potential unintended areas of impact

Feature being accessible in unintended builds of the app or for non wpcom sites.

2. What I did to test those areas of impact (or what existing automated tests I relied on)4. 

I manually tested the WordPress app and viewed other site types (e.g. self hosted, Jetpack), and verified the feature flag settings.

3. What automated tests I added (or what prevented me from doing so)

N/A

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by mokagio** [comment]
@frosty I'm going to push this to the 19.9 milestone. I think that's okay because this targets the ad hoc build, which makes me think there was no urgency to ship with 19.8. If that's not the case, let me know and I'll revert so this can go out with a beta as soon as approved 👍

**Comment by frosty** [comment]
@mokagio Correct, this is just for testing purposes at the moment – thanks!

File: WordPress/Classes/Utility/BuildInformation/FeatureFlag.swift [hunk 1]
```diff
@@ -62,7 +62,7 @@ enum FeatureFlag: Int, CaseIterable, OverrideableFlag {
         case .weeklyRoundupBGProcessingTask:
             return true
         case .domains:
-            return false
+            return BuildConfiguration.current ~= [.localDeveloper, .a8cBranchTest]
```
Comment (guarani) [inline WordPress/Classes/Utility/BuildInformation/FeatureFlag.swift]: 👋 This can now be updated to make Domains available on all Jetpack app builds in preparation for shipping in Jetpack 19.9.

**Review by guarani — COMMENTED** [review]
Hi @frosty, thanks for setting this up and apologies for the long delay testing this.

> Build and run the Jetpack target

❓I did most of my testing using the PR build. That doesn't make any difference, does it?

> View the Menu section of My Site for a number of sites – try to verify WordPress.com sites, self-hosted non-Jetpack and self-hosted Jetpack.

✅  I confirmed the Domain Dashboard (My Site → Domains) is shown for WP.com sites, and is hidden for self-hosted Jetpack sites. 

⚠️ I discovered that self-hosted non-Jetpack sites can't be accessed using the Jetpack app i.e. it's not possible to log in without installing the Jetpack plugin. Does this match your understanding?

<details>
  <summary>See screenshot</summary>
  
</details> 

> For WordPress.com sites, you should see the Domains item in the My Site menu, provided that you have site management permissions.

✅ I managed a WP.com site as an Editor and the Domain Dashboard was hidden.

> Navigate into Domains when it's present and ensure you can see the site's current domain listed.

✅  After purchasing a .blog domain (see details in next section), the Domains Dashboard showed the domain I'd purchased under the "YOUR SITE DOMAINS" section.

> Attempt the entire domains purchase flow. It's recommended to use the sandbox store – for full information about setting this up, read #p2-pbArwn-2GN. You should be able to successfully complete the flow, and the Domains Dashboard should update and reflect the new domain.

✅  After connecting the sandbox, I selected a .blog domain, entered dummy contact details and credit card information and completed the flow, seeing a congratulatory message at the end. 

❓I then returned to the Domains Dashboard and it says that my .blog domain "Never expires" – is that expected?

❓Afterwards, I disconnected the sandbox (logged out of the app and in again), and my .blog domain is still in the Domains Dashboard. I'm not sure if this is expected since it's unclear to me whether the domain should persist after the disconnection from the sandbox. Could you please confirm? Update: After visiting the Domains Dashboard some time later, the domain was no longer there. I'm still not sure what the correct behavior is.

> Also verify that the Domains section isn't visible in the WordPress app.

✅ I downloaded the WordPress build and the Domains Dashboard was hidden.

---

I noticed after tapping "Select domain", on the modal web view screen that's shown there is a "Purchase Details" dropdown. Tapping it reveals a "Questions? Read more about plans and purchases" link. That link is unresponsive when tapped.

  <details>
    <summary>See screen recording</summary>
     
  </details>

**Comment by guarani** [comment]
I left an approval on the accompanying WPKit PR: https://github.com/wordpress-mobile/WordPressKit-iOS/pull/497#pullrequestreview-970065734.

With regard to the testing required in order to be confident in launching Domains, are we relying on previous testing to cover the wide range of scenarios that users could encounter here (e.g. different payment types, different plan scenarios, localization, failure scenarios, edge cases, etc)? I ask because the testing instructions doesn't appear too comprehensive, but I understand that if testing has already been done to a sufficient level, we might not need to re-test all scenarios.

**Comment by frosty** [comment]
> ⚠️ I discovered that self-hosted non-Jetpack sites can't be accessed using the Jetpack app i.e. it's not possible to log in without installing the Jetpack plugin. Does this match your understanding?

This is correct – apologies for the misinformation in the testing steps!

>❓I then returned to the Domains Dashboard and it says that my .blog domain "Never expires" – is that expected?

Not expected. I've just observed this too. After a full refresh of the domains info, we do get a correct expiry / renewal date, so I suspect this means we need to perform a further refresh once everything's completed on the server side. I'll look at making a change to improve this behaviour.

> ❓Afterwards, I disconnected the sandbox (logged out of the app and in again), and my .blog domain is still in the Domains Dashboard. I'm not sure if this is expected since it's unclear to me whether the domain should persist after the disconnection from the sandbox. Could you please confirm? Update: After visiting the Domains Dashboard some time later, the domain was no longer there. I'm still not sure what the correct behavior is.

I've observed this too and I'm going to put it down to a quirk of the sandbox environment. I verified that the domain definitely hasn't been purchased for the user and doesn't show in their live purchases, only in the sandbox, yet we're still seeing it in the normal response for a while.

> I noticed after tapping "Select domain", on the modal web view screen that's shown there is a "Purchase Details" dropdown. Tapping it reveals a "Questions? Read more about plans and purchases" link. That link is unresponsive when tapped.

Navigation would be disabled here on purpose so that users can't access details of payments for plans etc via the app.

> With regard to the testing required in order to be confident in launching Domains, are we relying on previous testing to cover the wide range of scenarios that users could encounter here (e.g. different payment types, different plan scenarios, localization, failure scenarios, edge cases, etc)? 

The majority of the flow used here is web view based, using existing web payment flows that have been used and tested extensively. There is of course the creation of the cart and display of the web view, but as far as I'm aware there aren't too many options for how that's created so if we've tested the flow it should be covered 🤞 There's also the domain credits flow, but that has been exposed in the apps for some time already via a message on the user's My Sites dashboard, so there's not too much new going on there.

**Comment by frosty** [comment]
@Guarani Just pushed a new version:

- Opens up the feature flag to all Jetpack builds
- Uses the latest prerelease version of WordPressKit which is now available
-  Refresh the domains dashboard list after a successful purchase, which should fix the 'never expires' issue

**Comment by guarani** [comment]
> Navigation would be disabled here on purpose so that users can't access details of payments for plans etc via the app.

My concern is that users won't understand why the link doesn't work. Also, while link was about plans, the link seems to be dynamic. For example, I see "Questions? Ask a Happiness Engineer" on [the Android build](https://github.com/wordpress-mobile/WordPress-Android/pull/16417) now and it similarly doesn't work (cc @ravishanker):

<details>
  <summary>See WPAndroid screenshot</summary>
  Screen Shot 2022-05-12 at 19 14 51
</details>

**Comment by guarani** [comment]
After internal discussion ([see here](p1652456156075939/1652398028.045269-slack-C441E3YTS)), we're going to merge this with the disabled links, for inclusion in 19.9 which is cut on Monday.

**Review by guarani — APPROVED** [review]
Code looks good! Merging this now as per above comment.
